RNA-seq for Mendelian disease diagnostics: A hands-on tutorial through bioinformatic tools and workflows
===
An invitate RNA-seq workshop @ASHG2019
---

Welcome to the invitate ASHG RNA-seq workshop!

We setup a Slack workspace to facilitate communication within the participants as well as the rare disease RNA-seq community in general.
Please sign up [here](https://join.slack.com/t/rnaseq4rare/shared_invite/enQtNzU1MzEyOTI4NjQ3LTMwNjM1NmZlNWY3Nzk1MjkxYTFkYjBjNjdlNTI3Y2ZkNzJjZTNmZTFiZDVhOTVhMjEwYjRiYzA4Y2QwMzhjNTA). We hope to have vivid conversations and interactions during the workshop but also in the future within this workspace.

Through the workshop we will use interactive R notebooks (IR notebooks), which are similar to the well known IPython notebooks.
We have for each part of the workshop one notebook. In the end of each notebook there is always a little exercise where you can test your new skills. 

How to make the most out of this tutorial
---

To use this tutorial most efficiently, you should follow some rules:

1. Open the notebook in Google Colab (click on the Colab button on the top of the notebook)
1. Save a copy in your personal Google Drive (click on `File -> Save a copy in Drive ...`
1. Run the `r-env-setup-script.R` part to setup the R environment (takes ~3 minutes)
1. Have fun learning new skills!
1. Communicate with your peers in person (most important) or on Slack!

If you are new to programming in R or to notebooks in general have a look into the [getting started guide](colab_notebooks/Getting_Started_with_Colab_and_R.ipynb).

## The notebooks for the workshop

* [Expression Outlier Detection](colab_notebooks/Expression_Outlier_Detection.ipynb)
* [Aberrant Splicing Detection](TODO)
* [Mono Allelic Expression Analysis](TODO)
* [Variant Prioritization](TODO)

## Links to the presentations used in the workshop

* [Introduction]()
* [Expression Outlier Detection]()
* [Mono Allelic Expression Analysis]()
* [Variant Prioritization]()
* [Conclusion]()

