---
webDir: /s/public_webshare/project/ashg_rnaseq_workshop
projectTitle: ASHG19 RNA-seq Workshop
htmlOutputPath: Output/ashg19_rnaseq/html
fileRegex: .*\.R

PROJECT_NAME: ashg19_rnaseq_workschop

# OUTPUT Directory
ROOT: /s/project

# INPUT: Sample annotation
SAMPLE_FILE_MAPPING: /s/project/ashg19_rnaseq_workschop/Data/raw_data/filemapping.tsv
SAMPLE_ANNOTATION:   /s/project/ashg19_rnaseq_workschop/Data/raw_data/annotation.tsv

# Sample file mapping: assays
rna_assay: RNA_ID
dna_assay: VCF_ID

# INPUT: Gene Annotation
# names must match in GENE_ANNOTATION and in GENE_INFO
# used in aberrant_expression/Outrider/FilterCounts.R
GENE_ANNOTATION: ["/s/genomes/human/hg19/gencode29/gencode.v29lift37.annotation.gtf.gz"]
COUNT_RANGES:    ["tmp/exons_by_gene.Rds"]
GENE_ANNOTATION_NAMES: ["v29"]
GENE_INFO: ["tmp/gencode_v29_unique_gene_name.tsv"]

CHROMOSOME_FORMAT: UCSC

#### Aberrant Expression
outrider_group: OUTRIDER_GROUP
strand_column: COUNT_AS_STRANDED
inter_feature_column: COUNT_OVERLAP
min_outrider_ids: 40
fpkmCutoff: 1


#### Variant Annotation
gene_assembly: "hg19"
VEP_VERSION: 94
CHROMOSOME_NAMES: "1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,X,Y,MT"
filter_vcf: {"FILTER": ["PASS"]} ## Example: {"FILTER": ["PASS"], "QUAL" : 20} for numeric values it will be tested with >= 

